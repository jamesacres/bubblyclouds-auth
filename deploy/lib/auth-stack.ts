import { Duration, Fn, Stack, StackProps } from 'aws-cdk-lib';
import {
  CfnApplication,
  CfnConfigurationProfile,
  CfnEnvironment,
} from 'aws-cdk-lib/aws-appconfig';
import {
  DomainName,
  EndpointType,
  LambdaIntegration,
  MethodLoggingLevel,
  RestApi,
  SecurityPolicy,
} from 'aws-cdk-lib/aws-apigateway';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';
import { AttributeType, ProjectionType, Table } from 'aws-cdk-lib/aws-dynamodb';
import {
  LayerVersion,
  ParamsAndSecretsLayerVersion,
  ParamsAndSecretsLogLevel,
  ParamsAndSecretsVersions,
  Runtime,
} from 'aws-cdk-lib/aws-lambda';
import {
  NodejsFunction,
  NodejsFunctionProps,
} from 'aws-cdk-lib/aws-lambda-nodejs';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';
import { Secret } from 'aws-cdk-lib/aws-secretsmanager';
import { Construct } from 'constructs';
import * as path from 'path';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';

export interface AuthStackProps extends StackProps {
  certificateArn: string;
  domainName: string;
  subdomain: string;
  appConfig: { applicationName: string; environmentName: string };
}

export class AuthStack extends Stack {
  constructor(scope: Construct, id: string, props: AuthStackProps) {
    super(scope, id, props);
    const { certificateArn, domainName, subdomain } = props;

    const authGateway = this.gateway();
    this.domain(authGateway, {
      certificateArn,
      domainName,
      subdomain,
    });

    const appConfig = this.appConfig(props.appConfig);
    const { oidc, redirect } = this.lambdas({
      appConfig,
      accountId: props.env!.account!,
      region: props.env!.region!,
    });

    const { sigRSA } = this.secrets();
    sigRSA.grantRead(oidc.fn);

    const { table } = this.dynamodb();
    table.grantReadWriteData(oidc.fn);

    // GET /
    authGateway.root.addMethod('GET', redirect.integration);

    const jwksResource = authGateway.root.addResource('jwks');
    // GET /jwks
    jwksResource.addMethod('GET', oidc.integration);

    const wellKnownResource = authGateway.root.addResource('.well-known');
    const openidConfigurationResource = wellKnownResource.addResource(
      'openid-configuration'
    );
    // GET /.well-known/openid-configuration
    openidConfigurationResource.addMethod('GET', oidc.integration);

    const oidcResource = authGateway.root.addResource('oidc');
    // GET /oidc
    oidcResource.addMethod('GET', redirect.integration);

    // GET /oidc/jwks
    const oidcJwksResource = oidcResource.addResource('jwks');
    oidcJwksResource.addMethod('GET', oidc.integration);

    const authResource = oidcResource.addResource('auth');
    // GET /oidc/auth
    authResource.addMethod('GET', oidc.integration);
    const authUidResource = authResource.addResource('{uid}');
    // GET /oidc/auth/:uid
    authUidResource.addMethod('GET', oidc.integration);

    const tokenResource = oidcResource.addResource('token');
    // POST /oidc/token
    tokenResource.addMethod('POST', oidc.integration);

    const sessionResource = oidcResource.addResource('session');

    const sessionEndResource = sessionResource.addResource('end');
    // GET /oidc/session/end
    sessionEndResource.addMethod('GET', oidc.integration);

    const sessionEndConfirmResource = sessionEndResource.addResource('confirm');
    // POST /oidc/session/end/confirm
    sessionEndConfirmResource.addMethod('POST', oidc.integration);

    const sessionEndSuccessResource = sessionEndResource.addResource('success');
    // GET /oidc/session/end/success
    sessionEndSuccessResource.addMethod('GET', oidc.integration);

    const interactionResource = oidcResource.addResource('interaction');

    const interactionCallbackResource =
      interactionResource.addResource('callback');

    const interactionCallbackGoogleResource =
      interactionCallbackResource.addResource('google');
    // GET /oidc/interaction/callback/google
    interactionCallbackGoogleResource.addMethod('GET', oidc.integration);

    const interactionCallbackAppleResource =
      interactionCallbackResource.addResource('apple');
    // POST /oidc/interaction/callback/apple
    interactionCallbackAppleResource.addMethod('POST', oidc.integration);

    const interactionUidResource = interactionResource.addResource('{uid}');
    // GET /oidc/interaction/:uid
    interactionUidResource.addMethod('GET', oidc.integration);

    const interactionUidFederatedResource =
      interactionUidResource.addResource('federated');
    // POST /oidc/interaction/:uid/federated
    interactionUidFederatedResource.addMethod('POST', oidc.integration);

    const interactionUidFederatedGoogleResource =
      interactionUidFederatedResource.addResource('google');
    // GET /oidc/interaction/:uid/federated/google
    interactionUidFederatedGoogleResource.addMethod('GET', oidc.integration);

    const interactionUidFederatedAppleResource =
      interactionUidFederatedResource.addResource('apple');
    // GET /oidc/interaction/:uid/federated/apple
    interactionUidFederatedAppleResource.addMethod('GET', oidc.integration);

    const interactionUidConfirmResource =
      interactionUidResource.addResource('confirm');
    // POST /oidc/interaction/:uid/confirm
    interactionUidConfirmResource.addMethod('POST', oidc.integration);

    const interactionUidAbortResource =
      interactionUidResource.addResource('abort');
    // GET /oidc/interaction/:uid/abort
    interactionUidAbortResource.addMethod('GET', oidc.integration);
  }

  private gateway() {
    return new RestApi(this, 'AuthGateway', {
      restApiName: 'AuthGateway',
      deployOptions: {
        metricsEnabled: true,
        loggingLevel: MethodLoggingLevel.ERROR,
        tracingEnabled: true,
      },
      cloudWatchRole: true,
      endpointTypes: [EndpointType.EDGE],
    });
  }

  private domain(
    authGateway: RestApi,
    {
      certificateArn,
      domainName,
      subdomain,
    }: {
      certificateArn: string;
      domainName: string;
      subdomain: string;
    }
  ) {
    const certificate = Certificate.fromCertificateArn(
      this,
      'Certificate',
      certificateArn
    );
    const domain = new DomainName(this, 'AuthDomain', {
      certificate,
      domainName: `${subdomain}.${domainName}`,
      securityPolicy: SecurityPolicy.TLS_1_2,
      endpointType: EndpointType.EDGE,
    });
    domain.addBasePathMapping(authGateway);
  }

  private secrets() {
    return {
      sigRSA: new Secret(this, 'sigRSA', { secretName: 'sigRSA' }),
    };
  }

  private dynamodb() {
    const table = new Table(this, 'AuthTable', {
      partitionKey: { name: 'modelId', type: AttributeType.STRING },
      pointInTimeRecovery: true,
      timeToLiveAttribute: 'expiresAt',
      deletionProtection: true,
    });

    // As per src/adapters/dynamodb.ts define secondary indexes
    ['uid', 'grantId', 'userCode'].map((column) =>
      table.addGlobalSecondaryIndex({
        indexName: `${column}Index`,
        partitionKey: { name: column, type: AttributeType.STRING },
        projectionType: ProjectionType.ALL,
      })
    );

    return { table };
  }

  private appConfig(options: AuthStackProps['appConfig']): {
    application: CfnApplication;
    environment: CfnEnvironment;
    configuration: CfnConfigurationProfile;
  } {
    const application = new CfnApplication(this, `AuthAppConfigApplication`, {
      name: options.applicationName,
    });
    const environment = new CfnEnvironment(this, 'AuthAppConfigEnvironment', {
      applicationId: application.ref,
      name: options.environmentName,
    });
    const configuration = new CfnConfigurationProfile(
      this,
      `AuthAppConfigConfigurationProfile`,
      {
        applicationId: application.ref,
        locationUri: 'hosted',
        name: options.environmentName,
        type: 'AWS.Freeform',
      }
    );
    return {
      application,
      environment,
      configuration,
    };
  }

  private lambdas(options: {
    accountId: string;
    region: string;
    appConfig: {
      application: CfnApplication;
      environment: CfnEnvironment;
      configuration: CfnConfigurationProfile;
    };
  }) {
    const config: NodejsFunctionProps = {
      handler: 'handler',
      memorySize: 128,
      environment: {},
      runtime: Runtime.NODEJS_20_X,
      timeout: Duration.seconds(15),
      bundling: {
        externalModules: ['@aws-sdk/*'],
      },
      logRetention: RetentionDays.ONE_WEEK,
    };

    const redirectFn = new NodejsFunction(this, `AuthRedirectFunction`, {
      ...config,
      entry: path.resolve(__dirname, '../../src/handlers/redirect.ts'),
      functionName: `AuthRedirect`,
    });

    const paramsAndSecrets = ParamsAndSecretsLayerVersion.fromVersion(
      ParamsAndSecretsVersions.V1_0_103,
      { logLevel: ParamsAndSecretsLogLevel.DEBUG }
    );

    const oidcFn = new NodejsFunction(this, `AuthOidcFunction`, {
      ...config,
      paramsAndSecrets,
      memorySize: 512,
      entry: path.resolve(__dirname, '../../src/handlers/oidc.ts'),
      functionName: `AuthOidc`,
      environment: {
        DEBUG: 'oidc-provider:*',
        OAUTH_TABLE: 'AuthStack-AuthTable0711E62F-15KG9EHHEGFYW',
        // https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-integration-lambda-extensions.html
        AWS_APPCONFIG_EXTENSION_PREFETCH_LIST: Fn.sub(
          '/applications/${applicationId}/environments/${environmentId}/configurations/${configurationId}',
          {
            applicationId: options.appConfig.application.name,
            environmentId: options.appConfig.environment.name,
            configurationId: options.appConfig.configuration.name,
          }
        ),
      },
      layers: [
        LayerVersion.fromLayerVersionArn(
          this,
          'AppConfigLambdaLayer',
          'arn:aws:lambda:eu-west-2:282860088358:layer:AWS-AppConfig-Extension:93'
        ),
      ],
    });
    oidcFn.addToRolePolicy(
      new PolicyStatement({
        resources: [
          Fn.sub(
            'arn:aws:appconfig:${region}:${accountId}:application/${applicationId}/environment/${environmentId}/configuration/${configurationId}',
            {
              region: options.region,
              accountId: options.accountId,
              applicationId: options.appConfig.application.ref,
              environmentId: options.appConfig.environment.ref,
              configurationId: options.appConfig.configuration.ref,
            }
          ),
        ],
        actions: [
          'appconfig:GetLatestConfiguration',
          'appconfig:StartConfigurationSession',
        ],
      })
    );

    return {
      redirect: {
        fn: redirectFn,
        integration: new LambdaIntegration(redirectFn),
      },
      oidc: { fn: oidcFn, integration: new LambdaIntegration(oidcFn) },
    };
  }
}
