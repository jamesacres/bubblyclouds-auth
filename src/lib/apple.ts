import openid, { Client } from 'openid-client';

let appleClient: Client;
export const getAppleClient = async (
  clientId: string,
  clientSecret: string,
  redirectUrl: string
) => {
  if (!appleClient) {
    const apple = await openid.Issuer.discover(
      'https://appleid.apple.com/.well-known/openid-configuration'
    );
    appleClient = new apple.Client({
      client_id: clientId,
      client_secret: clientSecret,
      response_types: ['code'],
      redirect_uris: [redirectUrl],
      grant_types: ['authorization_code'],
    });
  }
  return appleClient;
};
