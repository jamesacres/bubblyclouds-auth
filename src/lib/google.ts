import openid, { Client } from 'openid-client';

let googleClient: Client;
export const getGoogleClient = async (
  clientId: string,
  redirectUrl: string
) => {
  if (!googleClient) {
    const google = await openid.Issuer.discover(
      'https://accounts.google.com/.well-known/openid-configuration'
    );
    googleClient = new google.Client({
      client_id: clientId,
      response_types: ['id_token'],
      redirect_uris: [redirectUrl],
      grant_types: ['implicit'],
    });
  }
  return googleClient;
};
