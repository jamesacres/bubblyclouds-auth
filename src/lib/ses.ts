import {
  SESv2Client,
  SendEmailCommand,
  SendEmailCommandInput,
} from '@aws-sdk/client-sesv2';

export class SesConfig {
  fromName: string;
  fromEmail: string;
  fromArn: string;
}

export class Ses {
  constructor(private config: SesConfig) {}

  sendEmail = async ({
    html,
    subject,
    text,
    toEmail,
  }: {
    html: string;
    subject: string;
    text: string;
    toEmail: string;
  }): Promise<void> => {
    const client = new SESv2Client();
    const input: SendEmailCommandInput = {
      FromEmailAddress: `${this.config.fromName} <${this.config.fromEmail}>`,
      FromEmailAddressIdentityArn: this.config.fromArn,
      Destination: {
        ToAddresses: [toEmail],
      },
      Content: {
        Simple: {
          Subject: {
            Data: subject,
          },
          Body: {
            Text: {
              Data: text,
            },
            Html: {
              Data: html,
            },
          },
        },
      },
    };
    const command = new SendEmailCommand(input);
    await client.send(command);
  };
}
