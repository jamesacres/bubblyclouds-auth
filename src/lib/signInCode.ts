import { DynamoDBAdapter } from '../adapters/dynamodb';
import { Model } from '../types/Model';
import { sanitiseEmail } from '../utils/email';
import { randomHumanCode } from '../utils/random';

export class SignInCode {
  private adapter = new DynamoDBAdapter(Model.BubblySignInCode);

  constructor() {}

  async getCode(requestEmail: string): Promise<string> {
    const email = sanitiseEmail(requestEmail);

    const existingCode = await this.adapter.find(email);
    if (existingCode && !existingCode.consumed && existingCode.signInCode) {
      return existingCode.signInCode;
    }

    const now = new Date();
    const expiresInSeconds = 3600; // one hour
    const signInCode = randomHumanCode();
    await this.adapter.upsert(
      email,
      {
        signInCode,
        createdAt: now.toISOString(),
        updatedAt: now.toISOString(),
      },
      expiresInSeconds
    );
    return signInCode;
  }

  async checkCode(requestEmail: string, requestCode: string): Promise<boolean> {
    const email = sanitiseEmail(requestEmail);
    if (
      typeof requestCode === 'string' &&
      requestCode.toLowerCase().replaceAll('-', '').trim() ===
        (await this.getCode(email)).toLowerCase().replaceAll('-', '').trim()
    ) {
      await this.adapter.consume(email);
      return true;
    }
    return false;
  }
}
