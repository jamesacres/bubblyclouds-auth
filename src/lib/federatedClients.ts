import { OAuth2Client as GoogleOAuth2Client } from 'google-auth-library';
import jwt from 'jsonwebtoken';
import { CallbackParamsType, Client, IdTokenClaims } from 'openid-client';
import { AppConfig } from '../types/AppConfig';
import { getAppleClient } from '../lib/apple';
import { getGoogleClient } from '../lib/google';
import { errors } from 'oidc-provider';

export interface FederatedClientsConfig {
  federatedClients: AppConfig['federatedClients'];
  serverUrl: string;
  serverUrlProd?: string;
}

export class FederatedClients {
  private _googleClient: Client;
  private _appleClient: Client;

  constructor(private config: FederatedClientsConfig) {}

  googleClient = async () => {
    if (!this._googleClient) {
      const {
        serverUrl,
        federatedClients: {
          google: { clientId },
        },
      } = this.config;
      const callbackUrl = `${serverUrl}/oidc/interaction/callback/google`;
      this._googleClient = await getGoogleClient(clientId, callbackUrl);
    }
    return this._googleClient;
  };

  googleIdTokenClaims = async (
    nonce: string,
    uid: string,
    callbackParams: CallbackParamsType
  ): Promise<IdTokenClaims> => {
    try {
      const tokenset = await (
        await this.googleClient()
      ).callback(undefined, callbackParams, {
        nonce,
        state: uid,
        response_type: 'id_token',
      });
      const googleOAuth2Client = new GoogleOAuth2Client();
      await googleOAuth2Client.verifyIdToken({
        idToken: tokenset.id_token!,
        audience: this.config.federatedClients.google.clientId,
      });
      return tokenset.claims();
    } catch (e) {
      console.error(e);
      throw new errors.InvalidToken('invalid google id_token');
    }
  };

  appleClient = async () => {
    if (!this._appleClient) {
      const {
        serverUrl,
        serverUrlProd,
        federatedClients: {
          apple: { teamId, clientId, privateKey, keyId },
        },
      } = this.config;
      const clientSecret = jwt.sign(
        {
          iss: teamId,
          aud: 'https://appleid.apple.com',
          sub: clientId,
        },
        privateKey,
        {
          algorithm: 'ES256',
          expiresIn: 15777000,
          header: {
            alg: 'ES256',
            kid: keyId,
          },
        }
      );
      const callbackUrl = `${serverUrlProd || serverUrl}/oidc/interaction/callback/apple`;
      this._appleClient = await getAppleClient(
        clientId,
        clientSecret,
        callbackUrl
      );
    }
    return this._appleClient;
  };

  appleIdTokenClaims = async (
    nonce: string,
    uid: string,
    callbackParams: CallbackParamsType
  ) => {
    try {
      const tokenset = await (
        await this.appleClient()
      ).callback(undefined, callbackParams, {
        nonce,
        state: uid,
        response_type: 'code',
      });
      return tokenset.claims();
    } catch (e) {
      console.error(e);
      throw new errors.InvalidToken('invalid apple id_token');
    }
  };
}
