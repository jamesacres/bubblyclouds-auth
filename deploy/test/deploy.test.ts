import { App } from 'aws-cdk-lib';
import { AuthStack } from '../lib/auth-stack';
import { Template } from 'aws-cdk-lib/assertions';

test('Auth Stack', () => {
  const app = new App();
  const authStack = new AuthStack(app, 'AuthStack', {
    env: {
      account: '12345678',
      region: 'eu-west-2',
    },
    certificateArn: 'mockCertificateArn',
    domainName: 'mockdomain.test',
    subdomain: 'mocksubdomain',
  });
  const template = Template.fromStack(authStack);
  expect(template.toJSON()).toMatchSnapshot();
});
