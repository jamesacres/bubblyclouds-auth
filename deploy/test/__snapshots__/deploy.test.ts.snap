// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Stack 1`] = `
{
  "Outputs": {
    "AuthGatewayEndpoint49671CC4": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "AuthGateway83D5F26D",
            },
            ".execute-api.eu-west-2.",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "AuthGatewayDeploymentStageprodAE1C2F72",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AuthDomain2370CC8A": {
      "Properties": {
        "CertificateArn": "mockCertificateArn",
        "DomainName": "mocksubdomain.mockdomain.test",
        "EndpointConfiguration": {
          "Types": [
            "EDGE",
          ],
        },
        "SecurityPolicy": "TLS_1_2",
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "AuthDomainMapAuthStackAuthGateway1EA582623C9FEA9B": {
      "Properties": {
        "DomainName": {
          "Ref": "AuthDomain2370CC8A",
        },
        "RestApiId": {
          "Ref": "AuthGateway83D5F26D",
        },
        "Stage": {
          "Ref": "AuthGatewayDeploymentStageprodAE1C2F72",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "AuthGateway83D5F26D": {
      "Properties": {
        "EndpointConfiguration": {
          "Types": [
            "EDGE",
          ],
        },
        "Name": "AuthGateway",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "AuthGatewayAccount17342C6A": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "AuthGateway83D5F26D",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "AuthGatewayCloudWatchRole465259DB",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "AuthGatewayCloudWatchRole465259DB": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "AuthGatewayDeploymentE35D8D60780d3829d041bb7166e4097bcfeed7bd": {
      "DependsOn": [
        "AuthGatewayoidcGETDE5F9E2F",
        "AuthGatewayoidc8A96E077",
      ],
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "AuthGateway83D5F26D",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "AuthGatewayDeploymentStageprodAE1C2F72": {
      "DependsOn": [
        "AuthGatewayAccount17342C6A",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "AuthGatewayDeploymentE35D8D60780d3829d041bb7166e4097bcfeed7bd",
        },
        "MethodSettings": [
          {
            "DataTraceEnabled": false,
            "HttpMethod": "*",
            "LoggingLevel": "ERROR",
            "MetricsEnabled": true,
            "ResourcePath": "/*",
          },
        ],
        "RestApiId": {
          "Ref": "AuthGateway83D5F26D",
        },
        "StageName": "prod",
        "TracingEnabled": true,
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "AuthGatewayoidc8A96E077": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "AuthGateway83D5F26D",
            "RootResourceId",
          ],
        },
        "PathPart": "oidc",
        "RestApiId": {
          "Ref": "AuthGateway83D5F26D",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "AuthGatewayoidcGETApiPermissionAuthStackAuthGateway1EA58262GEToidcC4F2019F": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "AuthOidcFunctionFEAD1639",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:eu-west-2:12345678:",
              {
                "Ref": "AuthGateway83D5F26D",
              },
              "/",
              {
                "Ref": "AuthGatewayDeploymentStageprodAE1C2F72",
              },
              "/GET/oidc",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "AuthGatewayoidcGETApiPermissionTestAuthStackAuthGateway1EA58262GEToidc009D1884": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "AuthOidcFunctionFEAD1639",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:eu-west-2:12345678:",
              {
                "Ref": "AuthGateway83D5F26D",
              },
              "/test-invoke-stage/GET/oidc",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "AuthGatewayoidcGETDE5F9E2F": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:eu-west-2:lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "AuthOidcFunctionFEAD1639",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "AuthGatewayoidc8A96E077",
        },
        "RestApiId": {
          "Ref": "AuthGateway83D5F26D",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "AuthOidcFunctionFEAD1639": {
      "DependsOn": [
        "AuthOidcFunctionServiceRole7F4533CF",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-12345678-eu-west-2",
          "S3Key": "a329291cf95cfd2c2842b319b32eb6f2c37d1e14e6d9096799be529ba1805304.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "FunctionName": "AuthOidc",
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "AuthOidcFunctionServiceRole7F4533CF",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AuthOidcFunctionLogRetentionC3B84C04": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "AuthOidcFunctionFEAD1639",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "AuthOidcFunctionServiceRole7F4533CF": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-12345678-eu-west-2",
          "S3Key": "4e26bf2d0a26f2097fb2b261f22bb51e3f6b4b52635777b1e54edbd8e2d58c35.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
